thrift/arg2/kv_iterator.go

```diff
@@ -0,0 +1,87 @@
+package arg2
+
+import "encoding/binary"
+
+// KeyValIterator is a iterator for reading tchannel-thrift Arg2 Scheme,
+// which has key/value pairs (k~2 v~2).
+// NOTE: to be optimized for performance, we try to limit the allocation
+// done in the process of iteration.
+type KeyValIterator struct {
+	arg2Len               int
+	leftPairCount         int
+	keyOffset, keyLen     int
+	valueOffset, valueLen int
+	arg2Payload           []byte
+}
+
+// InitKeyValIterator inits a KeyValIterator with the buffer pointing at
+// start of Arg2.
+// NOTE: tchannel-thrift Arg Scheme starts with number of key/value pair.
+func InitKeyValIterator(arg2Payload []byte, arg2Len int) (KeyValIterator, bool) {
+	if arg2Len < 2 {
+		return KeyValIterator{}, false
+	}
+
+	nh := int(binary.BigEndian.Uint16(arg2Payload[0:2]))
+	if nh <= 0 {
+		return KeyValIterator{}, false
+	}
+
+	return KeyValIterator{
+		leftPairCount: nh,
+		arg2Len:       arg2Len,
+		arg2Payload:   arg2Payload,
+	}.next(2 /*nh*/)
+}
+
+// Key Returns the key
+func (i KeyValIterator) Key() []byte {
+	return i.arg2Payload[i.keyOffset : i.keyOffset+i.keyLen]
+}
+
+// Value returns value.
+func (i KeyValIterator) Value() []byte {
+	return i.arg2Payload[i.valueOffset : i.valueOffset+i.valueLen]
+}
+
+// Next returns next iterator. Return nil if no more key/value pair available.
+func (i KeyValIterator) Next() (KeyValIterator, bool) {
+	if i.leftPairCount <= 0 {
+		return KeyValIterator{}, false
+	}
+
+	return i.next(i.valueOffset + i.valueLen)
+}
+
+// cur is the offset from the start of Arg2 payload to next key/value pair
+// we want to iterate to.
+func (i KeyValIterator) next(cur int) (KeyValIterator, bool) {
```

Comment:
hmm this is OK, but we're reimplementing what `typed.ReadBuffer` does, can we reuse some of that logic?

Reply:
The main reason I did not use it is typed.ReadBuffer is using allocation now.

Reply:
Ok, since the `relay_messages.go` doesn't use it either, let's leave that for later.